compiler/runtime/OMRCodeCache.hpp

```diff
@@ -52,29 +53,28 @@ extern uint8_t *align(uint8_t *ptr, uint32_t alignment);
 
 namespace OMR
 {
-
-// Base class capturing VM dependency on the _warmCodeAlloc and _coldCodeAlloc pointers at the
-// beginning of a code cache object. Do not change this base class without fully appreciating this
-// dependency.
+// Base class capturing VM dependency on the CodeCacheHeader containing 
+// warmCodeAlloc and coldCodeAlloc pointers at the beginning of a code cache object. 
+// Do not change this base class without fully appreciating this dependency.
 //
 class CodeCacheBase
    {
 public:
-   void setWarmCodeAlloc(uint8_t *wca)   { _warmCodeAlloc = wca; }
-   void setColdCodeAlloc(uint8_t *cca)   { _coldCodeAlloc = cca; }
+   void setWarmCodeAlloc(uint8_t *wca)   { _codeCacheHeader.warmCodeAlloc = wca; }
+   void setColdCodeAlloc(uint8_t *cca)   { _codeCacheHeader.coldCodeAlloc = cca; }
 
-   uint8_t *getWarmCodeAlloc()   { return _warmCodeAlloc; }
-   uint8_t *getColdCodeAlloc()   { return _coldCodeAlloc; }
+   uint8_t *getWarmCodeAlloc() const { return _codeCacheHeader.warmCodeAlloc; }
+   uint8_t *getColdCodeAlloc() const  { return _codeCacheHeader.coldCodeAlloc; }
 
-   void alignWarmCodeAlloc(uint32_t round)  { _warmCodeAlloc = align(_warmCodeAlloc, round); }
-   void alignColdCodeAlloc(uint32_t round)  { _coldCodeAlloc = align(_coldCodeAlloc, round); }
+   void alignWarmCodeAlloc(uint32_t round)  { _codeCacheHeader.warmCodeAlloc = align(_codeCacheHeader.warmCodeAlloc, round); }
+   void alignColdCodeAlloc(uint32_t round)  { _codeCacheHeader.coldCodeAlloc = align(_codeCacheHeader.coldCodeAlloc, round); }
 
-   // NOTE: warmCodeAlloc and coldCodeAlloc must be the first 2 entries
+   size_t getFreeContiguousSpace() const { return getColdCodeAlloc() - getWarmCodeAlloc(); }
+private:
+   // NOTE: CodeCacheHeader must be the first entry
```

Comment:
It would be nice if we could enforce this with a static assert.

Reply:
I tried the static_assert solution and here's what I found:
1) I cannot put the static_assert inside `CodeCacheBase` because the compiler wants to see the closing brace.
```
OMRCodeCache.hpp:78:18: error: invalid use of incomplete type ‘class OMR::CodeCacheBase’
    static_assert(offsetof(CodeCacheBase, _codeCacheHeader) == 0, "CodeCacheHeader must be the first entry of CodeCacheBase");
```
2) I can put the static_assert inside `class OMR_EXTENSIBLE CodeCache : public CodeCacheBase`, but then I need to make the new field `_codeCacheHeader` public. 

@Leonardo2718 Any suggestion?

Reply:
The field could be made protected instead of public. Or, the assert can be put in a member function (by convention, it could go in a constructor but CodeCacheBase doesn't have one explicitly defined).

I'd be ok with not having the assert if it overcomplicates things too much. However, I am a little concerned that the existing code is potentially relying on undefined behaviour. It's not clear to me yet whether this PR would resolve this or not.